---
title: "Assignment21"
author: "TB"
date: "31/01/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:



Assignment RNA-seq



Introduction questions:

Q1: Go online to the NASA GeneLab https://genelab-data.ndc.nasa.gov/genelab/accession/GLDS-38/ Read the Study Description and rephrase the scientific question in 20 words (~2 sentences)
Answer:



Q2: Approximately, how many plants were subjected to each condition? Search the information in the article https://bmcplantbiol.biomedcentral.com/articles/10.1186/s12870-020-02392-6 or online on the NASA GeneLab https://genelab-data.ndc.nasa.gov/genelab/accession/GLDS-38/
Answer:



Q3: What is the sequencing platform that was used?
Answer:



Q4: What are the softwares that were used to obtain the raw counts? List them with a small description of their role?  In particular, trimming, aligning and read counting are the most important.
Answer:


## Data import

Exercise 1

Practical Tasks Import the raw gene counts and the sample to conditions in your R workspace. 

```{r}
# import data



```


Scale the counts with DESeq2 into an R object called "scaled_counts" and save it as a .CSV file on your disk called "scaled_counts.csv".

```{r}
# Scale the data with the use of DESeq2


# Safe as .csv


```




Perform a few "sanity checks" (number of rows and columns, show the first 10 lines, etc.). You don't have to include this into your final report.  Make two graphs to compare the effect of the scaling:- 1 graph with the raw counts (1 boxplot per sample, 3 control and 3 spaceflight treatments)- 1 graph with the scaled counts  (1 boxplot per sample, 3 control and 3 spaceflight treatments)

```{r}
# Some sanity checks, like number of rows, collumns and visualize the first 10 lines


```

 Make two graphs to compare the effect of the scaling:- 1 graph with the raw counts (1 boxplot per sample, 3 control and 3 spaceflight treatments)- 1 graph with the scaled counts  (1 boxplot per sample, 3 control and 3 spaceflight treatments)
 
```{r}
# boxplot of the raw data.


```
 

```{r}
# boxplot of the scaled data.


```

Questions
Q5: according to you and based on the last figure, what is the goal of the scaling procedure? Write a small comment about the raw counts/scaled counts figures you have just created.
Answer:



##PCA analysis

Exercise 2: 
Principal Component Analysis (PCA)The PCA analysis will indicate how much variance can be explained by the first Principal Components (PCs) and these should overlap with the experimental design of the study.For the PCA, use the scaled counts "scaled_counts" and NOT the raw counts. 

Task: Compute the PCA as seen in the tutorial with the mypca() function:  https://scienceparkstudygroup.github.io/rna-seq-lesson/05-descriptive-plots/index.html#53-pca-applied-to-rna-seq, section 5.2. Save your PCA results into an object called "pca" in R that contains scores, loadings and explained variance.  

```{r}
# Compute the PCA as seen in the tutorial with the mypca() function


```




Make three plots:
Scree plot: explained variance per PC (only first 10 components are necessary).
```{r}
# Scree plot


```



Score plot: display the samples on PC1 and PC2. 
```{r}
# Score plot


```

Third plot???




Questions:
Q6: how much variance percentage is explained by PC1 and PC2? 
Answer:


Q7: Based on the score plot and the percentage of variance explained by PC1 and PC2, write a small comment on the quality of the experiment? Differential expression 
Answer:



##Differentially expressed genes

Task:
run the related DESeq2 steps to obtain differentially expressed genes (DE-genes). Call this R object "diff_genes". Task: display a table of the top 20 differentially expressed genes based on their significance value. 
```{r}
# get table of DE genes with DESeq2


# visualize the top 20


```


Questions

Q8: How many genes are differentially expressed (threshold of p = 0.01)?
Answer:



Q9: Among the top 20 diff. expressed genes, how many are up-regulated due to the lack of gravity?
Answer:



Q10: What are the maximum and minimum log2fold changes? Convert to a "regular" fold change (revert the log2)?      
Answer:



##Volcano plot

Practical tasks

Task: Make a vulcano plot, a type of scatterplot that shows statistical significance (P value) versus magnitude of change (fold change), as seen in the tutorial: https://scienceparkstudygroup.github.io/rna-seq-lesson/06-differential-analysis/index.html#3-volcano-plot section 3

```{r}
# "shrink" the log2 fold changes and store the result under "resLFC"


# load the library "EnhancedVolcano" and run the function


```




Questions:
Q11:
Where in the vulcano plot can you find: the most up-regulated, most down-regulated and the most statistically significant DE-genes
Answer:


##Adopt a gene

Task: Each of the two pair students choose one top DE-gene and add the required extra information by consulting external databases. 
Look for information on type of protein it codes for, what process is it involved in, where and when is it active, Does it help to answer to the scientific question as you wrote in question Q1?


