---
title: "Assignments"
---

<!-- MarkdownTOC autolink="true" -->

- [Dataset 1: Arabidopsis in Space \(Kruse 2020\)](#dataset-1-arabidopsis-in-space-kruse-2020)
	- [Abstract](#abstract)
	- [Dataset](#dataset)
	- [Publication](#publication)
- [Dataset 2: seed longevity \(Renard 2020\)](#dataset-2-seed-longevity-renard-2020)
	- [Abstract](#abstract-1)
	- [Dataset](#dataset-1)
	- [Publication](#publication-1)
- [Previous assignments based on Vogel 2016](#previous-assignments-based-on-vogel-2016)
	- [Assignment 1](#assignment-1)
	- [Assignment 2](#assignment-2)
	- [Assignment 3](#assignment-3)
	- [Assigment 4](#assigment-4)
	- [Tasks](#tasks)

<!-- /MarkdownTOC -->


Assignments are mainly designed for student curriculums but can be also used to test your working knowledge upon completion of the lesson. 

# Dataset 1: Arabidopsis in Space (Kruse 2020)

## Abstract
> On Earth plants are constantly exposed to a gravitational field of 1G. Gravity affects a plant in every step of its development. Germinating seedlings orient their radicle and hypocotyl and growing plants position organs at a specific Gravitropic Set-point Angle dictated by the asymmetric distribution of auxin depending on the gravity vector. Hence gravitropism is one of the fundamental growth responses in plants. For any experiment studying the effects of gravity on plants, the ultimate control is the microgravity in space. In this study, Arabidopsis seeds were flown to the International Space Station and allowed to germinate and grow for 3 days in microgravity. Arabidopsis Wild Type Col-0 seeds were plated onto twenty-two 60mm Petri plates, loaded into PDFUs and inserted 4 Biological Research in Canisters (BRICs). Approximately 800 seeds were sterilized, plated on each 60mm Petri plates and cold stratified for 16 hours followed by 2 hours of white light treatment. The BRICs were maintained at 4C until spaceflight to ensure seed germination in microgravity. After 3 days of germination and growth, the seedlings were fixed by injecting RNAlater into the chamber. They were kept at ambient temperature for 12 hours followed by freezing at -80C. An additional 22 plates were used as ground controls. After the spaceflight, tissue from five plates was pooled to make each of three replicates. Both membrane and soluble proteins were extracted from the pooled seedlings. Proteins were trypsin digested, labelled with iTRAQ and identified using tandem mass spectrometry. 

## Dataset

__Source: NASA GeneLab:__ [https://genelab-data.ndc.nasa.gov/genelab/accession/GLDS-38/](https://genelab-data.ndc.nasa.gov/genelab/accession/GLDS-38/)

Zenodo link to processed data = normalised counts. 

## Publication 

Kruse, C.P.S., Meyers, A.D., Basu, P. et al. Spaceflight induces novel regulatory responses in Arabidopsis seedling as revealed by combined proteomic and transcriptomic analyses. BMC Plant Biol 20, 237 (2020). [https://doi.org/10.1186/s12870-020-02392-6](https://doi.org/10.1186/s12870-020-02392-6)

# Dataset 2: seed longevity (Renard 2020)

## Abstract

## Dataset

## Publication
Renard, J., Martínez-Almonacid, I., Sonntag, A., Molina, I., Moya-Cuevas, J., Bissoli, G., Muñoz-Bertomeu, J., Faus, I., Niñoles, R., Shigeto, J., Tsutsumi, Y., Gadea, J., Serrano, R., & Bueso, E. (2020). PRX2 and PRX25, peroxidases regulated by COG1, are involved in seed longevity in Arabidopsis. Plant, cell & environment, 43(2), 315–326. https://doi.org/10.1111/pce.13656
Format:

# Previous assignments based on Vogel 2016
Each assigment comprises two comparisons (comp #1 and comp #2). 

## Assignment 1

| seed                	| infection            	| dpi 	| comparison 	| assignment 	|
|---------------------	|----------------------	|-----	|------------	|------------	|
| MgCl2                	| P. syringae DC3000   	| 2   	| comp #1    	| A1          |
| S. melonis Fr1     	  | P. syringae DC3000 	  | 2   	| comp #1     | A1         	|
| MgCl2               	| P. syringae DC3000   	| 7   	| comp #2    	| A1         	|
| S. melonis Fr1 	      | P. syringae DC3000 	  | 7   	| comp #2  	  | A1         	|

* Comparison 1: 2978 genes (p < 0.01)
* Comparison 2: 3031 genes (p < 0.01)

## Assignment 2

| seed                	| infection            	| dpi 	| comparison 	| assignment 	|
|---------------------	|----------------------	|-----	|------------	|------------	|
| MgCl2                	| Mock                	| 2     | comp #1     | A2     	  |
| MgCl2             	  | P. syringae DC3000    | 2     | comp #1     | A2         	|
| MgCl2                	| Mock               	  | 7     | comp #2  	  | A2         	|
| MgCl2         	      | P. syringae DC3000    | 7     | comp #2     | A2         	|

* Comparison 1: 795 genes (p < 0.01)
* Comparison 2: 3249 genes (__p < 0.001__) and around 4979 at p < 0.01 (comparison from the tutorial).

## Assignment 3

| seed                	| infection            	| dpi 	| comparison 	| assignment 	|
|---------------------	|----------------------	|-----	|------------	|------------	|
| M. extorquens PA1     | Mock               	  | 2   	| comp #1    	| A3        	|
| S. melonis Fr1        | Mock                  | 2   	| comp #1   	| A3          	|
| M. extorquens PA1     | Mock               	  | 7   	| comp #2    	| A3          	|
| S. melonis Fr1        | Mock              	  | 7   	| comp #2    	| A3          	|


* Comparison 1: 1008 genes (p < 0.01)
* Comparison 2: 680 genes (p < 0.01) 

## Assigment 4

| seed                	| infection            	| dpi 	| comparison 	| assignment 	|
|---------------------	|----------------------	|-----	|------------	|------------	|
| M. extorquens PA1     | P. syringae DC3000    | 2   	| comp #1    	| A4    	|
| S. melonis Fr1        | P. syringae DC3000    | 2   	| comp #1    	| A4          	|
| M. extorquens PA1     | P. syringae DC3000   	| 7   	| comp #2    	| A4          	|
| S. melonis Fr1        | P. syringae DC3000   	| 7   	| comp #2    	| A4          	|

* Comparison 1: 3189 genes (p < 0.01)
* Comparison 2: 696 genes (p < 0.01) 

## Tasks
1. Rephrase the scientific question (0.5 point).
2. Compute the tables of differential expressed genes for both comparisons. Create a histogram of corrected p-values for each comparison (QC). Annotate the genes with their gene symbols and annotation using biomartr (2 points). 
3. Create a volcano plot for each comparison to display the differential genes. Justify your choice for the fold change cutoff based on the distribution of your log2 fold change values (e.g. median, 75th percentile, etc). 


{% include links.md %}